src/Bridge.php

```diff
@@ -1,117 +1,137 @@
+        /** @var \Cake\Http\Response $response */
+        $response = (new Runner())->run($middleware, $request, $this->application);
+
         $cookies = [];
         foreach ($response->getCookieCollection() as $cookie) {
-            /**
-             * @var Cookie $cookie
-             */
-            if ($cookie->getExpiresTimestamp() === '0') {
+            if (!$cookie->getExpiresTimestamp()) {
                 $cookie = $cookie->withNeverExpire();
             }
             $cookies[] = $cookie->toHeaderValue();
         }
         if (!empty($cookies)) {
             $response = $response->withHeader('Set-Cookie', $cookies);
         }
-        if (!($response instanceof \Psr\Http\Message\ResponseInterface)) {
-            throw new \RuntimeException(sprintf(
-                'Application did not create a response. Got "%s" instead.',
-                is_object($response) ? get_class($response) : $response
-            ));
-        }
 
         session_write_close();
 
         return $response;
     }
 
-    protected function convertRequest(\Zend\Diactoros\ServerRequest $request) : \Cake\Http\ServerRequest
+    /**
+     * Convert ServerRequestInterface to Cake ServerRequest. This is necessary since some CakePHP internals and some
+     * plugin middleware require an instance of Cake ServerRequest.
+     *
+     * @todo result of `$request->getParsedBody()` is always null, see link tag below this todo. We rely on
+     *      BodyParserMiddleware after setting the body on the ServerRequest below.
+     * @link https://github.com/roadrunner-server/roadrunner/discussions/953
+     * @param \Psr\Http\Message\ServerRequestInterface $request An instance of ServerRequestInterface
+     * @return \Cake\Http\ServerRequest
+     */
+    private function convertRequest(ServerRequestInterface $request): ServerRequest
     {
-        $server = $request->getServerParams();
-        $server['REQUEST_TIME'] = time();
-        $server['REQUEST_TIME_FLOAT'] = microtime(true);
-        $server['REMOTE_ADDR'] = '127.0.0.1';
-        $server['SERVER_PROTOCOL'] = $request->getUri()->getScheme();
-        $server['REQUEST_METHOD'] = $request->getMethod();
-        $server['SERVER_NAME'] = $request->getUri()->getHost();
-        $server['SERVER_PORT'] = $request->getUri()->getPort();
-        $server['REQUEST_URI'] = $request->getUri()->getPath();
-
-        $query = $request->getQueryParams();
-        $body = $request->getParsedBody();
-        $cookies = $request->getCookieParams();
-        $files = $request->getUploadedFiles();
-
-        $cakeRequest = ServerRequestFactory::fromGlobals($server, $query, $body, $cookies, $files);
+        $cakeRequest = ServerRequestFactory::fromGlobals(
+            $request->getServerParams(),
+            $request->getQueryParams(),
+            $request->getParsedBody(),
+            $request->getCookieParams(),
+            $request->getUploadedFiles()
+        );
         $cakeRequest->trustProxy = true;
```

Comment:
Why is this needed @steinkel ?

Reply:
The main reason was providing translation between \Zend\Diactoros\ServerRequest into \Cake\Http\ServerRequest, it was not possible to pass the request without conversion from/to the cakephp stack for processing without this utility in those times...

Reply:
Specifically I mean the `$cakeRequest->trustProxy = true;`

Reply:
I barely remember it was needed to correctly pass the correct IP to cake context, as the go server was acting as a proxy